### ISSUE TITLE
🧪 A/B Testing Framework — ML Shield Validation (Phase 8) 7.4.2026

### ISSUE THREAD

**Issue by VaclavSercl** [VaclavSercl/sniper#20]
### Přehled
Systém pro objektivní měření, zda ML Shield (Fáze 6) skutečně zlepšuje obchodní výsledky. Porovnává dva simultánní běhy: s ML biasem (treatment) a bez něj (control), a vyhodnocuje statistickou signifikanci rozdílu v PnL, fill rate, a toxic flow.

### Motivace
- ML Shield běží v produkci od v18.0, ale nemáme **kvantitativní důkaz** že pomáhá
- Bez A/B testingu riskujeme:
  - ML model, který přidává šum místo hodnoty
  - Falešný pocit bezpečí z "AI" bez měřitelného zlepšení
  - Nemožnost optimalizovat hyperparametry (learning rate, horizonsize)
- **Cíl**: Statisticky prokázat, že ML Shield zlepšuje Sharpe ratio o ≥ 0.1

---

### Architektura

```
┌──────────────────────────────────────────┐
│           A/B Testing Controller         │
│         (architect/ab_testing.py)         │
├──────────────────────────────────────────┤
│                                          │
│  Session A (Control)     Session B (ML)  │
│  ┌──────────────────┐   ┌──────────────┐│
│  │ ML Shield OFF    │   │ ML Shield ON ││
│  │ skew_adj = 0     │   │ skew_adj = X ││
│  │ confidence = 0   │   │ confidence=Y ││
│  └──────────────────┘   └──────────────┘│
│           │                     │        │
│           ▼                     ▼        │
│     PnL Tracking          PnL Tracking   │
│     Fill Analysis         Fill Analysis  │
│           │                     │        │
│           └─────────┬──────────┘        │
│                     ▼                    │
│           Statistical Comparison         │
│           (t-test, Sharpe ratio)          │
└──────────────────────────────────────────┘
```

---

#### 1. Session Manager
- **Alternating mode**: Každou hodinu přepíná ML Shield ON/OFF
  - 0:00–0:59 = ML ON
  - 1:00–1:59 = ML OFF
  - 2:00–2:59 = ML ON
  - ...
- **Control**: Zapisuje `l1_skew_adjustment = 0` a `l1_confidence_score = 0` do engine_state.bin mmap v OFF periodách
- **Logging**: Každý fill je tagován `session_id` a `ml_enabled: true/false`

#### 2. Metrics Collector
Pro každou session (ON vs OFF) sbírá:

| Metrika | Popis | Proč je důležitá |
|---------|-------|-------------------|
| **Realized PnL** | FIFO PnL per session | Hlavní úspěšnostní metrika |
| **Fill count** | Počet vyplněných objednávek | ML by neměl snížit aktivitu |
| **Toxic fill rate** | % fillů, kde cena ihned šla proti nám | ML by měl snížit toxic fills |
| **Avg spread captured** | Průměrný realizovaný spread | ML by měl zlepšit spread capture |
| **Sharpe ratio** | PnL / stdev(PnL) per session | Risk-adjusted výkon |
| **Win rate** | % profitabilních fillů | Konzistence |
| **Max drawdown** | Největší propad | ML by neměl zvýšit riziko |

#### 3. Statistical Analysis
- **Welchův t-test**: Porovnání průměrů ON vs OFF
- **Minimum sample size**: 100 fillů per group (≈ 1-3 dny)
- **Confidence level**: 95% (p < 0.05)
- **Effect size**: Cohenovo d > 0.2 (small effect)

#### 4. Reporting
- **Telegram report**: `/ab` command — aktuální stav A/B testu
- **Automatický weekly report**: Shrnutí po 7 dnech testování
- **Dashboard panel**: Live A/B comparison v master_dashboard.html

---

### Data Storage

```sql
-- Nová tabulka v pnl.db
CREATE TABLE ab_sessions (
    id INTEGER PRIMARY KEY,
    start_ts TEXT NOT NULL,
    end_ts TEXT,
    ml_enabled INTEGER NOT NULL,  -- 0=control, 1=treatment
    fills INTEGER DEFAULT 0,
    realized_pnl REAL DEFAULT 0,
    toxic_fills INTEGER DEFAULT 0,
    avg_spread_bps REAL DEFAULT 0,
    max_drawdown REAL DEFAULT 0
);

CREATE TABLE ab_fills (
    id INTEGER PRIMARY KEY,
    session_id INTEGER REFERENCES ab_sessions(id),
    timestamp TEXT NOT NULL,
    side TEXT NOT NULL,           -- buy/sell
    price REAL NOT NULL,
    amount REAL NOT NULL,
    pnl REAL DEFAULT 0,
    is_toxic INTEGER DEFAULT 0,  -- 1 if price moved against within 5s
    ml_skew REAL,                -- ML bias at time of fill
    ml_confidence REAL           -- ML confidence at time of fill
);
```

---

### Implementace

```
architect/
├── ab_testing.py       — Session controller, metrics collector, reporter
└── ab_analysis.py      — Statistical analysis, t-test, visualization
```

#### `ab_testing.py` — Session Controller
- Daemon thread v `tg_commander.py` (jako L2 Oracle loop)
- Hodiny sudé = ML ON, hodiny liché = ML OFF
- Zapisuje do `engine_state.bin` mmap (pole `l1_skew_adjustment`)
- Loguje fills z `pnl.db` do `ab_sessions`/`ab_fills`

#### `ab_analysis.py` — Statistical Analyzer
- Spouštěn jako cron nebo manuálně (`/ab` TG command)
- Čte `ab_sessions` + `ab_fills`
- Počítá t-test, Sharpe, effect size
- Generuje textový/markdown report

---

### Prerekvizity
- [x] ML Shield běží v produkci (`architect/ml_shield.py`)
- [x] PnL database (`shared/pnl_engine.py`)
- [x] Engine state mmap s ML fieldy (offset 1584, 1600)
- [ ] Minimum 1 týden produkčních dat pro statistickou signifikanci

### Rizika
- **Market regime změny** mohou zkreslit výsledky (ML ON v trending, ML OFF v ranging)
  - **Mitigace**: Randomizovat ON/OFF pořadí, logovat regime per session
- **Sample size**: Pokud Hydra dělá < 50 fillů/den, potřebujeme 2+ týdny
- **Confounding variables**: Jiné parametry se mohou měnit (L2 Oracle tuning)
  - **Mitigace**: Lockovat L2 Oracle parametry během A/B testu

---

### Fáze implementace
1. **Schema** — SQLite tabulky `ab_sessions`, `ab_fills`
2. **Controller** — Hodinové přepínání ML ON/OFF
3. **Collector** — Tagování fillů, PnL per session
4. **Analysis** — t-test, Sharpe, report generace
5. **Integration** — `/ab` TG command, dashboard panel
6. **Run** — Minimum 7 dní, vyhodnocení

### Priorita: P2 (important for validation, but not revenue-generating)
